articles/2016-10-12-functor-adts.txt

```diff
@@ -1,37 +1,38 @@
 package[value="articles"]{}
 
-post[id=functor-adts,date=2016-10-12,description="Using contravariance, type parameters and type members to implement a functor in ADTs",Implementing functors in ADTs] {
+post[id=functor-adts,date=2016-10-12,description="Review of Scala language features for call site type polymorphism",Call site type polymorphism] {
   section[Motivation] {
-    It is a common problem to define url[https://en.wikipedia.org/wiki/Algebraic_data_type]{Algebraic Data Types} (ADTs). To prevent duplication, shared functionality should be ideally kept in base code{trait}s. Oftentimes, a problem arises that in the base code{trait} we do not know the child type the function is called on.
+    It is a common problem to define class hierarchies. To prevent duplication, shared functionality should be ideally kept in base code{trait}s. Oftentimes, a problem arises that in the base code{trait} we do not know the child type the function is called on:
 
-    In other words, we want to be able to implement operations such as code{map()} in terms of a functor:
-    scala[type=section,file=FunctorAdts,value=functor]{}
+    scala[type=section,file=FunctorAdts,value=problem]{}
 
-    In this article, I am going to outline different ways how to solve this particular problem in Scala.
+    In other words, for a class code{F[_]} we want to define methods in its base code{trait}. These methods should have the return type code{F[_]}. In this article, I am going to outline different ways how to solve this particular problem in Scala.
   }
 
   section[Contravariance] {
     Consider a simple binary tree:
     scala[type=section,file=FunctorAdts,value=bin-tree]{}
 
-    We are going to leverage the fact that functions are contravariant in their return types. Let us extend our ADT by a code{map()} function:
+    We are going to leverage the fact that functions are contravariant in their return types. Let us extend our hierarchy by a code{map()} function:
```

Comment:
The functions are co-variant in their return types and contra-variant in their input types: http://www.scala-lang.org/api/current/index.html#scala.Function1